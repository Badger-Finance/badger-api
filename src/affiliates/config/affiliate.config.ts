import { BadRequest } from '@tsed/exceptions';
import { Chain } from '../../chains/config/chain.config';
import { Protocol } from '../../config/constants';
import { SettAffiliateData } from '../../setts/interfaces/sett-affiliate-data.interface';
import { SettDefinition } from '../../setts/interfaces/sett-definition.interface';

type Affiliates = { [protocol: string]: Affiliate };

export abstract class Affiliate {
  private static affiliates: Affiliates = {};

  readonly name: string;

  constructor(name: string) {
    this.name = name;
  }

  static register(protocol: Protocol, affiliate: Affiliate): void {
    this.affiliates[protocol] = affiliate;
  }

  static getAffiliate(affiliateData: SettAffiliateData): Affiliate {
    const affiliate = this.affiliates[affiliateData.protocol];
    if (!affiliate) {
      throw new BadRequest(`${affiliate} is not a affiliate`);
    }
    return affiliate;
  }

  abstract getAffiliateVaultData(chain: Chain, sett: SettDefinition): Promise<SettAffiliateData>;
}
