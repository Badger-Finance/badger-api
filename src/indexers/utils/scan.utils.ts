import { Network } from '@badger-dao/sdk';
import { BlocksRangeOptions } from '@badger-dao/sdk/lib/common';

import { getDataMapper } from '../../aws/dynamodb.utils';
import { IndexingMetadata } from '../../aws/models/indexing-metadata.model';
import { MAX_SCAN_RANGE } from '../constants/scan.constants';
import { LastScannedBlockMeta } from '../interfaces/last-scanned-block-meta.interface';

export async function getOrCreateMetadata<T>(task: string): Promise<IndexingMetadata<T>> {
  const mapper = getDataMapper();

  let metadata!: IndexingMetadata<T>;

  try {
    metadata = await mapper.get(Object.assign(new IndexingMetadata<T>(), { task }));
  } catch (e) {
    metadata = Object.assign(new IndexingMetadata<T>(), { task, data: {} });
  }

  return metadata;
}

export function getLastScanedBlock<T extends LastScannedBlockMeta>(
  indexingMetaData: T,
  network: Network,
  address: string,
) {
  return indexingMetaData[network]?.[address]?.lastScannedBlock || 0;
}

export function getScanRangeOpts(
  fromChain: Required<BlocksRangeOptions>,
  lastScannedBlock: number,
): Required<BlocksRangeOptions> {
  const startBlock = lastScannedBlock > 0 ? lastScannedBlock + 1 : fromChain.startBlock;
  let endBlock = fromChain.endBlock;

  if (endBlock - startBlock > MAX_SCAN_RANGE) {
    endBlock = startBlock + MAX_SCAN_RANGE;
  }

  return {
    startBlock,
    endBlock,
  };
}

export async function saveIndexingMetadata<T>(model: IndexingMetadata<T>): Promise<void> {
  const mapper = getDataMapper();

  try {
    await mapper.put(model);
  } catch (e) {
    console.error(`Failed to save indexing metadata for ${model.task}`, e);
  }
}
