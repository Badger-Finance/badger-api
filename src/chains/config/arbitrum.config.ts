import { Protocol } from '../../config/enums/protocol.enum';
import { SettState } from '../../config/enums/sett-state.enum';
import { Stage } from '../../config/enums/stage.enum';
import rpc from '../../config/rpc.config';
import { TOKENS } from '../../config/tokens.config';
import { GasPrices } from '../../gas/interfaces/gas-prices.interface';
import { getCurveSettTokenBalance } from '../../protocols/strategies/convex.strategy';
import { BouncerType } from '../../rewards/enums/bouncer-type.enum';
import { SettDefinition } from '../../setts/interfaces/sett-definition.interface';
import { arbitrumTokensConfig } from '../../tokens/config/arbitrum-tokens.config';
import { ChainNetwork } from '../enums/chain-network.enum';
import { ArbitrumStrategy } from '../strategies/arbitrum.strategy';
import { Chain } from './chain.config';

export class Arbitrum extends Chain {
  constructor() {
    super(
      'Arbitrum',
      'arbitrum',
      '0xa4b1',
      ChainNetwork.Arbitrum,
      arbitrumTokensConfig,
      arbitrumSetts,
      rpc[ChainNetwork.Arbitrum],
      new ArbitrumStrategy(Object.keys(arbitrumTokensConfig)),
      2425847,
      '0x635EB2C39C75954bb53Ebc011BDC6AfAAcE115A6',
      '0x85E1cACAe9a63429394d68Db59E14af74143c61c',
    );
    Chain.register(this.network, this);
  }

  async getGasPrices(): Promise<GasPrices> {
    const gasPrice = 2;
    return {
      rapid: gasPrice,
      fast: gasPrice,
      standard: gasPrice,
      slow: gasPrice,
    };
  }
}

export const arbitrumSetts: SettDefinition[] = [
  {
    name: 'Sushiswap Wrapped Ether/Sushi Helper',
    settToken: TOKENS.BARB_SUSHI_WETH_SUSHI,
    depositToken: TOKENS.ARB_SUSHI_WETH_SUSHI,
    createdBlock: 13163959,
    protocol: Protocol.Sushiswap,
    strategy: '0x86f772C82914f5bFD168f99e208d0FC2C371e9C2',
  },
  {
    name: 'Sushiswap Wrapped BTC/Wrapped ETH',
    settToken: TOKENS.BARB_SUSHI_WETH_WBTC,
    depositToken: TOKENS.ARB_SUSHI_WETH_WBTC,
    createdBlock: 13163959,
    protocol: Protocol.Sushiswap,
    strategy: '0xA6827f0f14D0B83dB925B616d820434697328c22',
  },
  {
    name: 'Curve renBTC/wBTC',
    settToken: TOKENS.BARB_CRV_RENBTC,
    depositToken: TOKENS.ARB_CRV_RENBTC,
    getTokenBalance: getCurveSettTokenBalance,
    createdBlock: 13237551,
    protocol: Protocol.Curve,
  },
  {
    name: 'Curve Tricrypto',
    settToken: TOKENS.BARB_CRV_TRICRYPTO,
    depositToken: TOKENS.ARB_CRV_TRICRYPTO,
    getTokenBalance: getCurveSettTokenBalance,
    createdBlock: 13237551,
    protocol: Protocol.Curve,
  },
  {
    name: 'Curve Tricrypto Light',
    settToken: TOKENS.BARB_CRV_TRICRYPTO_LITE,
    depositToken: TOKENS.ARB_CRV_TRICRYPTO,
    getTokenBalance: getCurveSettTokenBalance,
    createdBlock: 13321375,
    stage: Stage.Staging,
    state: SettState.Experimental,
    protocol: Protocol.Curve,
  },
  {
    name: 'Swapr Swapr/Wrapped ETH',
    settToken: TOKENS.BARB_SWP_SWPR_WETH,
    depositToken: TOKENS.ARB_SWP_SWPR_WETH,
    createdBlock: 13315350,
    state: SettState.Guarded,
    protocol: Protocol.Swapr,
  },
  {
    name: 'Swapr Wrapped BTC/Wrapped ETH',
    settToken: TOKENS.BARB_SWP_WBTC_WETH,
    depositToken: TOKENS.ARB_SWP_WBTC_WETH,
    createdBlock: 13315350,
    state: SettState.Guarded,
    protocol: Protocol.Swapr,
  },
  {
    name: 'Swapr Badger/Wrapped ETH',
    bouncer: BouncerType.Badger,
    settToken: TOKENS.BARB_SWP_BADGER_WETH,
    depositToken: TOKENS.ARB_SWP_BADGER_WETH,
    createdBlock: 2188169,
    state: SettState.Guarded,
    protocol: Protocol.Swapr,
  },
];
