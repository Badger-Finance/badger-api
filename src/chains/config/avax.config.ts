import { Network, Protocol, VaultState, VaultVersion } from '@badger-dao/sdk';
import rpc from '../../config/rpc.config';
import { TOKENS } from '../../config/tokens.config';
import { GasPrices } from '../../gas/interfaces/gas-prices.interface';
import { VaultDefinition } from '../../vaults/interfaces/vault-definition.interface';
import { Chain } from './chain.config';
import { BaseStrategy } from '../strategies/base.strategy';
import { avalancheTokensConfig } from '../../tokens/config/avax-tokens.config';
import { Stage } from '../../config/enums/stage.enum';

export class Avalanche extends Chain {
  constructor() {
    super(
      'Avalanche',
      'avalanche',
      '0xa86a',
      Network.Avalanche,
      avalancheTokensConfig,
      rpc[Network.Avalanche],
      avalancheSetts,
      new BaseStrategy(Network.Avalanche, Object.keys(avalancheTokensConfig)),
    );
    Chain.register(this.network, this);
  }

  async getGasPrices(): Promise<GasPrices> {
    return this.defaultGasPrice();
  }

  // TODO: Update if badger is ever updated
  getBadgerTokenAddress(): string {
    return TOKENS.BADGER;
  }
}

export const avalancheSetts: VaultDefinition[] = [
  {
    name: 'WBTC',
    vaultToken: TOKENS.BAVAX_WBTC,
    depositToken: TOKENS.AVAX_WBTC,
    protocol: Protocol.Aave,
    stage: Stage.Staging,
    state: VaultState.Experimental,
    version: VaultVersion.v1_5,
  },
];
