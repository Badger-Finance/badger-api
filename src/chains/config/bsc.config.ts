import { Network } from "@badger-dao/sdk";

import rpc from "../../config/rpc.config";
import { TOKENS } from "../../config/tokens.config";
import { bscTokensConfig } from "../../tokens/config/bsc-tokens.config";
import { BaseStrategy } from "../strategies/base.strategy";
import { Chain } from "./chain.config";

export class BinanceSmartChain extends Chain {
  constructor() {
    super(Network.BinanceSmartChain, bscTokensConfig, rpc[Network.BinanceSmartChain], new BaseStrategy(Network.BinanceSmartChain));
    Chain.register(this.network, this);
  }

  getBadgerTokenAddress(): string {
    return TOKENS.MULTI_BADGER;
  }
}
