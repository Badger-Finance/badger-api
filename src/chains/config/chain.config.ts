import { BadRequest } from '@tsed/exceptions';
import { ethers } from 'ethers';
import { STAGE } from '../../config/constants';
import { GasPrices } from '../../gas/interfaces/gas-prices.interface';
import { SettDefinition } from '../../setts/interfaces/sett-definition.interface';
import { TokenConfig } from '../../tokens/interfaces/token-config.interface';
import { ChainStrategy } from '../strategies/chain.strategy';
import { Network } from '@badger-dao/sdk';

type Chains = Record<string, Chain>;

export abstract class Chain {
  private static chains: Chains = {};
  readonly name: string;
  readonly symbol: string;
  readonly chainId: string;
  readonly network: Network;
  readonly tokens: TokenConfig;
  readonly setts: SettDefinition[];
  readonly provider: ethers.providers.JsonRpcProvider;
  readonly batchProvider: ethers.providers.JsonRpcBatchProvider;
  readonly strategy: ChainStrategy;
  readonly graphUrl: string;
  readonly blocksPerYear: number;
  readonly badgerTree?: string;
  readonly rewardsLogger?: string;

  constructor(
    name: string,
    symbol: string,
    chainId: string,
    network: Network,
    tokens: TokenConfig,
    setts: SettDefinition[],
    rpcUrl: string,
    strategy: ChainStrategy,
    blocksPerYear: number,
    badgerTree?: string,
    rewardsLogger?: string,
  ) {
    this.name = name;
    this.symbol = symbol;
    this.chainId = chainId;
    this.network = network;
    this.tokens = tokens;
    this.setts = setts.filter((sett) => !sett.stage || sett.stage === STAGE);
    this.provider = new ethers.providers.JsonRpcProvider(rpcUrl);
    this.batchProvider = new ethers.providers.JsonRpcBatchProvider(rpcUrl);
    this.strategy = strategy;
    this.graphUrl = `https://api.thegraph.com/subgraphs/name/badger-finance/badger-dao-setts${
      network !== Network.Ethereum ? `-${symbol.toLowerCase()}` : ''
    }`;
    this.blocksPerYear = blocksPerYear;
    this.badgerTree = badgerTree;
    this.rewardsLogger = rewardsLogger;
  }

  static register(network: Network, chain: Chain): void {
    Chain.chains[network] = chain;
    Chain.chains[chain.symbol] = chain;
    if (network === Network.Polygon) {
      Chain.chains['matic'] = chain;
    }
  }

  static getChain(network?: Network): Chain {
    if (!network) {
      network = Network.Ethereum;
    }
    const chain = this.chains[network];
    if (!chain) {
      throw new BadRequest(`${network} is not a supported chain`);
    }
    return chain;
  }

  abstract getGasPrices(): Promise<GasPrices>;
}
