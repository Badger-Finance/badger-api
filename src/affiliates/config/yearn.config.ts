import { BigNumber, ethers } from 'ethers';
import { Chain } from '../../chains/config/chain.config';
import { guestListAbi } from '../../config/abi/guest-list.abi';
import { yearnAffiliateVaultWrapperAbi } from '../../config/abi/yearn-affiliate-vault-wrapper.abi';
import { Protocol } from '../../config/constants';
import { SettAffiliateData } from '../../setts/interfaces/sett-affiliate-data.interface';
import { SettDefinition } from '../../setts/interfaces/sett-definition.interface';
import { getToken } from '../../tokens/tokens.utils';
import { Affiliate } from './affiliate.config';

export class Yearn extends Affiliate {
  constructor() {
    super(Protocol.Yearn);
    Affiliate.register(Protocol.Yearn, this);
  }

  async getAffiliateVaultData(chain: Chain, sett: SettDefinition): Promise<SettAffiliateData> {
    const affiliateData: SettAffiliateData = {
      protocol: Protocol.Yearn,
    };
    if (sett.hasBouncer) {
      try {
        const vaultWrapper = new ethers.Contract(sett.settToken, yearnAffiliateVaultWrapperAbi, chain.provider);
        const guestListAddress = await vaultWrapper.guestList();
        const guesList = new ethers.Contract(guestListAddress, guestListAbi, chain.provider);
        const [available, limit]: [BigNumber, BigNumber] = await Promise.all([
          guesList.remainingTotalDepositAllowed(),
          guesList.totalDepositCap(),
        ]);
        const depositToken = getToken(sett.depositToken);
        affiliateData.availableDepositLimit = parseFloat(ethers.utils.formatUnits(available, depositToken.decimals));
        affiliateData.depositLimit = parseFloat(ethers.utils.formatUnits(limit, depositToken.decimals));
      } catch {
        affiliateData.availableDepositLimit = 0;
        affiliateData.depositLimit = 50;
      }
    }
    return affiliateData;
  }
}
