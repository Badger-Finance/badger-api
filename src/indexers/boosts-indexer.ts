import { UnprocessableEntity } from '@tsed/exceptions';
import { ethers } from 'ethers';
import { getBoostFile } from '../accounts/accounts.utils';
import { CachedAccount } from '../aws/models/cached-account.model';
import { getDataMapper } from '../aws/dynamodb.utils';
import { Chain } from '../chains/config/chain.config';
import { BoostData } from '../rewards/interfaces/boost-data.interface';
import { CachedBoostMultiplier } from '../rewards/interfaces/cached-boost-multiplier.interface';
import { SUPPORTED_CHAINS } from '../chains/chain';

/**
 * Top level indexer call to separate chain updates.
 * Each chain is independently updated in its own promise
 * such that a single chain failure will not cause others to fail.
 */
export async function refreshUserBoosts() {
  await Promise.all(SUPPORTED_CHAINS.map((c) => exports.updateChainBoosts(c)));

  return 'done';
}

/**
 * Update user boost multipliers for a given chain.
 * @param chain Network users to update boost.
 */
export async function updateChainBoosts(chain: Chain): Promise<void> {
  try {
    // boosted data is only relevant on chains that have an emisison control
    if (!chain.emissionControl) {
      return;
    }

    // load the boost file, this is a requirement to be present for any chain with emisison control
    const file = await getBoostFile(chain);
    if (!file) {
      throw new UnprocessableEntity(`Unable to load boost file for ${chain.name}, expected file to exist!`);
    }

    console.log({
      chain: chain.network,
      multipliers: file.multiplierData,
    });
    const boostMultipliers = evaluateUserBoosts(chain, file);

    // construct batch queries for all users accounts, and fetch their stored entities
    const mapper = getDataMapper();
    const requestedAccounts = Object.keys(boostMultipliers).map((addr) =>
      Object.assign(new CachedAccount(), { address: ethers.utils.getAddress(addr) }),
    );
    const cachedAccounts = [];
    for await (const account of mapper.batchGet(requestedAccounts)) {
      cachedAccounts.push(account);
    }

    // update the stored account multipliers, removing old chain multipliers
    for (const account of cachedAccounts) {
      const userChainMultipliers = boostMultipliers[account.address];
      account.multipliers = account.multipliers.filter((m) => m.network !== chain.network).concat(userChainMultipliers);
    }

    for await (const _saved of mapper.batchPut(cachedAccounts));
  } catch (err) {
    console.error(err);
  }
}

/**
 * Evaluate the user boost and multipliers to update the user boost multipliers if required.
 * User boost multipliers supplied from the rewards JSON are the default values, and if the calculated
 * multiplier does not match (i.e. their boost has updated since last calculation) then the new estimate
 * will be displayed. This allows for a much shorter update time on the 'My Boost' values.
 * @param file Boost file to calculate user
 * @returns Mapping of user addresses to their new updated boost multipliers.
 */
export function evaluateUserBoosts(chain: Chain, file: BoostData): Record<string, CachedBoostMultiplier[]> {
  const { multiplierData, userData } = file;
  const boostMultipliers: Record<string, CachedBoostMultiplier[]> = {};

  /**
   * Iterate over all values in the boost file checking boost and multipliers.
   * Using the multiplier data from the boost file and the updated boost values
   * estimate the users current mutltiplier.
   *
   * If the keys do not match, or if the key is missing, this new estimated
   * boost multiplier will be set for the user.
   */
  for (const address of Object.keys(userData)) {
    const { multipliers, boost } = userData[address];
    const percentile = boost / 2000;
    Object.entries(multiplierData).forEach((entry) => {
      const [key, value] = entry;
      const range = value.max - value.min;
      const estimatedMultiplier = value.min + percentile * range;
      if (multipliers[key] !== estimatedMultiplier) {
        multipliers[key] = estimatedMultiplier;
      }
    });
    boostMultipliers[address] = Object.entries(multipliers).map((m) =>
      Object.assign(new CachedBoostMultiplier(), {
        address: m[0],
        network: chain.network,
        multiplier: m[1],
      }),
    );
  }

  return boostMultipliers;
}
