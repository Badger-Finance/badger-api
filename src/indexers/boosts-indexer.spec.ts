import { TOKENS } from '../config/tokens.config';
import { BoostMultiplierData } from '../rewards/interfaces/boost-multiplier-data.interface';
import { UserBoosts } from '../rewards/interfaces/user-boosts.interface';
import { TEST_ADDR, TEST_CHAIN } from '../test/tests.utils';
import * as boostIndexer from './boosts-indexer';

describe('boosts-indexer', () => {
  // common test data
  let multiplierData: BoostMultiplierData;
  let userData: UserBoosts;

  beforeEach(() => {
    multiplierData = {
      [TEST_ADDR]: {
        min: 1,
        max: 10,
      },
    };
    userData = {
      [TEST_ADDR]: {
        address: TEST_ADDR,
        rank: 30,
        boost: 1000,
        nativeBalance: 10,
        nonNativeBalance: 20,
        stakeRatio: 0.5,
        nftBalance: 10,
        bveCvxBalance: 5,
        diggBalance: 1,
        multipliers: {
          [TEST_ADDR]: 5.5,
        },
      },
    };
  });

  describe('evaluateUserBoosts', () => {
    it('returns an empty record on an empty boost file', () => {
      const result = boostIndexer.evaluateUserBoosts(TEST_CHAIN, {
        multiplierData: {},
        userData: {},
      });
      expect(result).toMatchObject({});
    });

    it('uses existing user boost if applicable', () => {
      const result = boostIndexer.evaluateUserBoosts(TEST_CHAIN, {
        multiplierData,
        userData,
      });
      expect(result).toMatchSnapshot();
    });

    it('updates user boost if applicable', () => {
      userData[TEST_ADDR].boost = 2000;
      const result = boostIndexer.evaluateUserBoosts(TEST_CHAIN, {
        multiplierData,
        userData,
      });
      expect(result).toMatchSnapshot();
    });

    it('adds user boost if applicable', () => {
      multiplierData[TOKENS.BBADGER] = { min: 2, max: 22 };
      const result = boostIndexer.evaluateUserBoosts(TEST_CHAIN, {
        multiplierData,
        userData,
      });
      expect(result).toMatchSnapshot();
    });
  });
});
