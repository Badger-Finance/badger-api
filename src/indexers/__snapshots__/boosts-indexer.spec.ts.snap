// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`boosts-indexer evaluateUserBoosts adds user boost if applicable 1`] = `
Object {
  "0xe6487033F5C8e2b4726AF54CA1449FEC18Bd1484": Array [
    CachedBoostMultiplier {
      "address": "0xe6487033F5C8e2b4726AF54CA1449FEC18Bd1484",
      "multiplier": 5.5,
      "network": "ethereum",
    },
    CachedBoostMultiplier {
      "address": "0x19D97D8fA813EE2f51aD4B4e04EA08bAf4DFfC28",
      "multiplier": 12,
      "network": "ethereum",
    },
  ],
}
`;

exports[`boosts-indexer evaluateUserBoosts updates user boost if applicable 1`] = `
Object {
  "0xe6487033F5C8e2b4726AF54CA1449FEC18Bd1484": Array [
    CachedBoostMultiplier {
      "address": "0xe6487033F5C8e2b4726AF54CA1449FEC18Bd1484",
      "multiplier": 10,
      "network": "ethereum",
    },
  ],
}
`;

exports[`boosts-indexer evaluateUserBoosts uses existing user boost if applicable 1`] = `
Object {
  "0xe6487033F5C8e2b4726AF54CA1449FEC18Bd1484": Array [
    CachedBoostMultiplier {
      "address": "0xe6487033F5C8e2b4726AF54CA1449FEC18Bd1484",
      "multiplier": 5.5,
      "network": "ethereum",
    },
  ],
}
`;
